docs/articles/documentation/recipes/best-practices/best-practices.md

```diff
@@ -0,0 +1,332 @@
+
+TestCafe is a tool built for functional testing. Do not use it to perform non-functional testing (like performance or load testing). Such tests would not yield any conclusive results.
+
+In your end-to-end tests, try to replicate real user's actions. Don't test exceptions - these are better tested with unit and integration tests. Test the general business logic of your application and abstain from rare use cases and edge scenarios.
+
+Write less E2E tests. End-to-end tests are slow by nature, their amount should be drastically lower than that of unit or integration tests.
+
+## Smart Assertions
+
+In end-to-end web testing, unpredictable factors (like network lag, processor, or memory bottlenecks in containers) can interfere with the assertions and produce inconsistent test results. Such tests are inconclusive (sometimes called 'flaky').
+
+TestCafe includes a [Smart Assertion Query Mechanism](../../guides/basic-guides/assert.md#smart-assertion-query-mechanism). This mechanism introduces wait time for all the assertions: if an assertion fails, it retries multiple times within a timeout. That reduces random impact and stabilizes the tests without performance trade-offs.
+
+The following example demonstrates a common mistake:
+
+```js
+import { Selector } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with Selector', async t => {
+    const developerNameInput = Selector('#developer-name');
+
+    await t.typeText(developerNameInput, 'Peter');
+
+    //an awaited selector doesn't update and produces unstable test results. Avoid it.
+    const developerName = await Selector('#developer-name').value;
+    //
+
+    await t
+            .expect(developerName).eql('Peter')
+            .typeText(developerNameInput, 'Jack')
+            .expect(developerName).eql('Jack'); // fails
+});
+```
+
+In this snippet, the `developerName` is initialized with the value of a Selector, but because of the `await` keyword, the value is calculated once and doesn't update. This disables the smart assertions query mechanism, which in this case leads to inconclusive results.
+
+To avoid this, omit the `await` keyword:
+
+```js
+import { Selector } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with Selector', async t => {
+    const developerNameInput = Selector('#developer-name');
+
+    await t.typeText(developerNameInput, 'Peter');
+
+    const developerName = Selector('#developer-name').value;
+
+    await t
+            .expect(developerName).eql('Peter')
+            .typeText(developerNameInput, 'Jack')
+            .expect(developerName).eql('Jack'); // passes
+});
+```
+
+The `developerName` is now initialized with a re-executable Selector API promise. When a test controller receives this promise, it enables the smart assertion query mechanism and TestCafe can wait for the value to update.
```

Comment:
> it enables the smart assertion query mechanism and TestCafe can wait for the value to update.

Boris always says that connecting different parts of a sentence with 'and' feels a bit Russian, and I tend to agree.

*it enables ... so TestCafe can wait ...*?

Reply:
> The `developerName` is now initialized with a re-executable Selector API promise. When a test controller receives this promise, it enables the smart assertion query mechanism so TestCafe can wait for the value to update.